Add docker capability

tudoramariei opened the pull request:
add Dockerfile, docker-compose, and dockerignore
add instructions in README

redecs commented on seismic-app/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+# base image
+FROM node:12.2.0-alpine
```
@tudoramariei do we need version pinning to be so specific? Or we can use use something like `node:12-apline`. This should gives the stable v12 api + minor changes / improvements (with BC) + security fixes.

redecs commented on seismic-app/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+ENV PATH /app/node_modules/.bin:$PATH
+
+# install and cache app dependencies
+COPY package.json /app/package.json
```
@tudoramariei You need to copy both `package.json` and `package-lock.json`, to make sure everyone is working on the same versions and we don't have any surprises (not everyone is using SemVer for their packages).

tudoramariei replied:
(not so) fun enough, if I add `COPY package.json /app/package-lock.json`, it seems to crash with the following error:
```log
ERROR: Service 'seismic-app' failed to build: The command '/bin/sh -c npm install' returned a non-zero code: 1
```

redecs replied:
@tudoramariei that's to be expected since you are basically renaming `package.json` to `package-lock.json` inside the container.
Try this: ```COPY package*.json ./```

tudoramariei replied:
OK, I saw the typo now 😅

redecs commented on seismic-app/docker-compose.yml:
```diff
@@ -0,0 +1,16 @@
+services:
+
+  sample:
+    container_name: sample
```
@tudoramariei I usually leave this out, since the the container name will be autogenerated from the current folder name + service name (when you use compose to manage multiple containers). Oh, this should be `seismic-app` in this context, not `sample`.

tudoramariei replied:
I'm guessing the `sample` from line 5 should also be changed, shouldn't it?

redecs replied:
Yes, that's right.

redecs reviewed (changes requested):
Looks good, but some tweaks required.

tudoramariei commented on seismic-app/README.md:
```diff
@@ -37,6 +37,12 @@ Instead, it will copy all the configuration files and the transitive dependencie
 
 You don’t have to ever use `eject`. The curated feature set is suitable for small and middle deployments, and you shouldn’t feel obligated to use this feature. However we understand that this tool wouldn’t be useful if you couldn’t customize it when you are ready for it.
 
+### `docker-compose up -d --build`
```
@redecs could you give me a tip here?
I'm not sure if this is the command people should use each time they need to start the container

redecs replied:
The command it's correct and it's the safe version.
Now for the nuances:
1. `-d` sends everything to background - I prefer it that way, but some people prefer to run it in foreground so they can see the logs. I prefer to use `docker-compose logs -f <service_name>`.
2. `--build` it's implied if you never started the container(s) before, but it's a good idea to include it every time, since this will rebuild the images (parts of it more specifically), if there are changes in the `Dockerfile`. If you leave this out you can start your project faster (when you already have the images built), but you might have an outdated setup, since you might be running a image based on an older `Dockerfile`.

tudoramariei replied:
OK, I'll leave it this way, then. Thanks for the explanation!

redecs commented on seismic-app/docker-compose.yml:
```diff
@@ -0,0 +1,16 @@
+      dockerfile: Dockerfile
+    volumes:
+      - '.:/app'
+      - '/app/node_modules'
```
@tudoramariei Just noticed this... It's OK to mount `node_modules` as a volume, especially when using Docker for Mac or Windows, but this means that you'll not have the node_modules folder on your machine, so you are loosing IDE support for those libraries. There is a way to configure this and have node_modules synced with the host machine, but it will incur a performance penalty on Windows (smalish) or MacOS (bigger).
So the questions is what do we want: IDE support for libraries or a bit more performance?

tudoramariei replied:
Most likely we're going to go for IDE support for the development part
It's also probably going to be useful to set multiple Dockerfiles `docker-compose` for deploying to production, I guess that's where I'd like to keep that line

redecs replied:
I'm in the same boat with IDE support. In production, volumes should be used only for data, not code.

redecs commented:
@tudoramariei I put an example together: https://gist.github.com/redecs/fa50b3d7f5ab44740af81c69856e0c9b 
Add `docker-entrypoint.sh` in the root of the project and adjust `Dockerfile` accordingly (lines   15-18 from my example `Dockerfile`).

tudoramariei commented:
I tried working with the gist but I still arrive at the same results. I just get `localhost refused to connect`

tudoramariei commented on seismic-app/Dockerfile:
```diff
@@ -9,8 +9,13 @@ ENV PATH /app/node_modules/.bin:$PATH
-RUN npm install
-RUN npm install react-scripts@3.2.0 -g
+
+COPY . ./
+
+COPY docker-entrypoint.sh /usr/local/bin/docker-entrypoint
+RUN chmod +x /usr/local/bin/docker-entrypoint
+
+ENTRYPOINT ["docker-entrypoint"]
```
@redecs These are the changes from the gist
Docker runs everything and I get a very nice `Starting seismic-app ... done` at the end, but I still can't connect to the app

redecs replied:
OK, I pull your branch of the code and I'll do some troubleshooting

tudoramariei replied:
Thanks a lot! 🥇

redecs replied:
@tudoramariei I got it to run. You just have to wait for the `npm ci` and `react-scripts start` commands to finish. For me, after docker-compose told the the container is up and running (`docker-compose ps`) it took another 2 minutes (Mac OS, older Mac). 
<img width="760" alt="image" src="https://user-images.githubusercontent.com/153277/69433929-3fa67700-0d45-11ea-88b8-6a02a8671394.png">

tudoramariei replied:
One more issue I see here (also on a Mac, 2016 one)
The container seems to die in ~5~ less than 2 seconds after it starts on my computer...

tudoramariei replied:
@redecs I found the problem. When looking at the docker logs for that container I saw that it crashes at `npm ci`.  I ran `npm install ci` on my machine and after 2 minutes the container is still alive 🙌 
I think adding `RUN npm install ci` in `Dockerfile` or `npm install ci` in `docker-entrypoint.sh` would be the solution. Which one would you recommend?

redecs replied:
`npm ci` it's a native npm command since 5.7,  and it should work without installing anything else. You might have some issues with that since you already have a `node_modules` directory generated by running npm command directly on the host system (Mac in your case).
I used `npm ci` instead of the regular `npm install` because I don't `npm install` to change the lock file (`packages-lock.json`). Usually I use `yarn` which has a flag special for this.
Here we could just replace `npm ci` with `npm install` and be done with it.

tudoramariei replied:
`npm install` seems to do the job just right
It's time to clean up the PR and get it ready for a `merge`

tudoramariei committed:
Dockerize application

Add the functionality to start the app through docker
- add docker ignore/compose/entrypoint/file
- add instructions in readme
